src/Servers/IIS/IIS/src/Core/IO/AsyncIOEngine.cs

```diff
@@ -35,13 +37,96 @@ public ValueTask<int> ReadAsync(Memory<byte> memory)
     }
 
     public ValueTask<int> WriteAsync(ReadOnlySequence<byte> data)
+    {
+        if (SegmentsOverChunksLimit(data))
+        {
+            return WriteDataOverChunksLimit(data);
+        }
+
+        return WriteDataAsync(data);
+    }
+
+    private ValueTask<int> WriteDataAsync(in ReadOnlySequence<byte> data)
     {
         var write = GetWriteOperation();
         write.Initialize(_handler, data);
         Run(write);
         return new ValueTask<int>(write, 0);
     }
 
+    /*
+    * In case the number of chunks is bigger than responseMaxChunks we need to make multiple calls 
+    * to the native api https://docs.microsoft.com/en-us/iis/web-development-reference/native-code-api-reference/ihttpresponse-writeentitychunks-method  
+    * Despite the documentation states that feeding the function with more than 65535 chunks will cause the function to throw an exception, 
+    * it actually seems that 65534 is the maximum number of chunks allowed.
+    * Also, there seems to be a problem when slicing a ReadOnlySequence on segment borders tracked here https://github.com/dotnet/runtime/issues/67607
+    * That's why we only allow 65533 chunks.  
+    */
+    private async ValueTask<int> WriteDataOverChunksLimit(ReadOnlySequence<byte> data)
+    {
+        var hr = 0;
+        ushort segmentsCount = 0;
+        var start = 0;
+        var length = 0;
+        int result;
+
+        foreach (var segment in data)
+        {
+            segmentsCount++;
+            length += segment.Length;
+
+            if (segmentsCount == responseMaxChunks)
+            {
+                result = await WriteDataAsync(data.Slice(start, length));
+
+                SaveResultIfError(result, ref hr);
+
+                segmentsCount = 0;
+                start = length;
+                length = 0;
+            }
+        }
+
+        if (segmentsCount > 0)
+        {
+            result = await WriteDataAsync(data.Slice(start, length));
+
+            SaveResultIfError(result, ref hr);
+        }
+
+        return hr;
+
+        static void SaveResultIfError(int result, ref int hr)
+        {
+            if (result > 0)
+            {
+                hr = result;
```

Comment:
Should have definitely short circuited, but I have still a doubt about this part, because I'm not sure the result we get is actually the hr from the native function, signaling success or error. In particular the `AsyncIOOperation.Complete` method, assign a different semantic to the result, based on the value of hr. 

Tell me if I'm wrong, but I feel like we are treating this value wrongly.

Reply:
We probably are handling this wrong. Upon further thought, I don't think we should short-circuit here considering we already ignore this HRESULT where it's surfaced and continue writing to the response body.

https://github.com/dotnet/aspnetcore/blob/6797f926dc6991294fb5e84505f13b672c2d97c6/src/Servers/IIS/IIS/src/Core/IISHttpContext.IO.cs#L172-L175

Can we just add a comment there that we're ignoring the HRESULT in the section of IISHttpContext.IO.cs I just referenced above, and mention if we start observing the HRESULT in `WriteDataOverChunksLimit` as well.

Let's get rid of `SaveResultIfError` and just return the last HRESULT.

Reply:
yes, I do agree.